src/types/slice.rs

```diff
@@ -42,31 +44,81 @@ impl PySliceIndices {
 }
 
 impl PySlice {
-    /// Constructs a new slice with the given elements.
+    /// Deprecated form of `PySlice::new_bound`.
+    #[track_caller]
+    #[cfg_attr(
+        not(feature = "gil-refs"),
+        deprecated(
+            since = "0.21.0",
+            note = "`PySlice::new` will be replaced by `PySlice::new_bound` in a future PyO3 version"
+        )
+    )]
     pub fn new(py: Python<'_>, start: isize, stop: isize, step: isize) -> &PySlice {
+        Self::new_bound(py, start, stop, step).into_gil_ref()
+    }
+
+    /// Constructs a new slice with the given elements.
+    pub fn new_bound(py: Python<'_>, start: isize, stop: isize, step: isize) -> Bound<'_, PySlice> {
         unsafe {
-            let ptr = ffi::PySlice_New(
+            ffi::PySlice_New(
                 ffi::PyLong_FromSsize_t(start),
                 ffi::PyLong_FromSsize_t(stop),
                 ffi::PyLong_FromSsize_t(step),
-            );
-            py.from_owned_ptr(ptr)
+            )
+            .assume_owned(py)
+            .downcast_into_unchecked()
         }
     }
 
-    /// Constructs a new full slice that is equivalent to `::`.
+    /// Deprecated form of `PySlice::full_bound`.
+    #[track_caller]
```

Comment:
I think we don't generally expect this API to panic so we probably don't need `track_caller` here.

```suggestion
```

Reply:
I guess this comes from where ever I took the deprecation warning from, haven looked close enough 🙃